src/mem/cache/prefetch/ipcp.cc

```diff
@@ -0,0 +1,398 @@
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * Describes a IPCP prefetcher based on template policies.
+ */
+
+#include "mem/cache/prefetch/ipcp.hh"
+
+#include "debug/HWPrefetch.hh"
+#include "params/IpcpPrefetcher.hh"
+
+namespace gem5
+{
+
+namespace prefetch
+{
+
+Ipcp::Ipcp(const IpcpPrefetcherParams &p)
+    : Queued(p), bitsIpTag(p.ip_tag_bits),
+    bitsLastVpage(p.last_vpage_bits),
+    bitsSignature(p.signature_bits),
+    degree(p.degree), csOn(p.cs_on),
+    cplxOn(p.cplx_on),
+    //GS IP Prefetcher
+    RM_SIZE(p.rm_size), REGION_SIZE(p.region_size),
+    pageDegree(p.page_degree), pageOn(p.page_on),
+    ipcpStats(this), nlOn(p.nl_on)
+{
+    bitsIpIndex = floorLog2(p.ip_entries);
+    bitsCsptIndex = floorLog2(p.cspt_entries);
+
+    std::cout << "bitsIpIndex: " << bitsIpIndex << std::endl;
+
+    ipcpTable.clear();
+    cspt.clear();
+
+    //GS IP Prefetcher
+    RST.clear();
+
+    misses = 0;
+    mpkc = 0;
+
+}
+
+Addr
+Ipcp::getSignature(Addr curSignature, int64_t stride)
+{
+    //Handle -ve stride
+    int64_t corStride = (stride < 0) ?
+     (stride * (-1)
+     + (1 << floorLog2(blkSize))) : stride;
+    return (((curSignature << 1) ^ corStride)
+     & ((1 << bitsSignature) - 1));
+}
+
+void
+Ipcp::calculatePrefetch(const PrefetchInfo &pfi,
+    std::vector<AddrPriority> &addresses,
+    const CacheAccessor &cache)
+{
+    if (!pfi.hasPC()) {
+        DPRINTF(HWPrefetch, "Ignoring request with no PC.\n");
+        return;
+    }
+
+    ipcpStats.cplxEntries = cspt.size();
+
+    //Calculate Misses per kilo cycles for NL
+    if (pfi.isCacheMiss()) misses += 1;
+    if ((ticksToCycles(curTick()) % 1000) == 0) {
+        mpkc = misses/1000;
+        misses = 0;
+    }
+
+    //Extract information
+    Addr pc = pfi.getPC();
+    Addr blkAddr = blockAddress(pfi.getAddr());
+    Addr regionBase = roundDown(blkAddr, REGION_SIZE);
+    Addr lineOffset = (blkAddr >> floorLog2(blkSize))
+     & (pageBytes/blkSize - 1);
+    //original index
+    Addr ipIndex = pc & ((1 << bitsIpIndex) - 1);
+
+    //Ideal
+    //Addr ipIndex = pc;
+    //Stride Index
+    //const Addr hash1 = pc >> 1;
+    //const Addr hash2 = hash1 >> (pc >> bitsIpIndex);
+    //Addr ipIndex = (hash1 ^ hash2) & ((1 << bitsIpIndex) - 1);
```

Comment:
Why is this commented out? If it's documentation, then it needs more context for future people to understand. If it's not documentation, it should be removed.